### DISCUSSION TITLE
Help with Progressive SSR rendering

### DISCUSSION THREAD

**Discussion by dgoh-zippia** [vercel/next.js#17435] [Help]
Hi all, currently my app is using SSR where all the data is fetched in NodeJS. I am trying to improve the First Paint and First Contentful Paint especially on low bandwidth network as the whole page needs to be built. I understand that SSR is when the server has to prepare the HTML before sending it to the user. However, I was wondering if it is possible to render in chunks when the data is ready. I read about stream SSR with renderToNodeStream but as far as I understand, NextJS doesn't support it. All help is appreciated.

**Comment by lfades** [comment]
Consider doing SSR at build-time, aka [Static Generation](https://nextjs.org/docs/basic-features/pages#static-generation-recommended) with [`getStaticProps`](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation). It will **always** be faster than HTML generated per request (SSR at runtime).

**Reply by dgoh-zippia** [reply]
Yeah I am trying to implement that as well. However, I am having trouble fetching the data. Should I do all the data fetching in getStaticProps() or do it in NodeJS and somehow pass it to getStaticProps function. If the latter is possible, how do I achieve it? Thank you.

**Reply by lfades** [reply]
> Should I do all the data fetching in getStaticProps() or do it in NodeJS

`getStaticProps` runs in Node.js 👍 . You should fetch the data that your visitors should see as soon as they enter the page, and if it's the same data for all of them. E.g blog posts.

**Reply by dgoh-zippia** [reply]
I see. Sorry to ask but can you give me an example of what that would look like? I am pretty new to this and I appreciate the help you have been giving me.

**Reply by enochndika** [reply]
Just read the doc. [data fetching](https://nextjs.org/docs/basic-features/data-fetching)

**Reply by dgoh-zippia** [reply]
I already read the doc but it didn't help me if I have a custom server as I am not sure how to pass the data to the page.

**Reply by enochndika** [reply]
i am curious why are you using a custom server?

**Reply by dgoh-zippia** [reply]
I mean app was built before NextJS released this feature and I want to do a variety of checks such as rate limiting

**Comment by ErnestBrandi** [comment]
> I was wondering if it is possible to render in chunks when the data is ready.

Sound more like client side rendering to me.
You could use [SWR](https://github.com/vercel/swr) which is also developed by the Vercel team.
